zebra-chain/src/orchard/commitment.rs

```diff
@@ -0,0 +1,428 @@
+
+/// LEBS2OSP256(repr_P(cv))
+///
+/// https://zips.z.cash/protocol/nu5.pdf#pallasandvesta
+impl From<ValueCommitment> for [u8; 32] {
+    fn from(cm: ValueCommitment) -> [u8; 32] {
+        cm.0.to_bytes()
+    }
+}
+
+impl<'a> std::ops::Sub<&'a ValueCommitment> for ValueCommitment {
+    type Output = Self;
+
+    fn sub(self, rhs: &'a ValueCommitment) -> Self::Output {
+        self - *rhs
+    }
+}
+
+impl std::ops::Sub<ValueCommitment> for ValueCommitment {
+    type Output = Self;
+
+    fn sub(self, rhs: ValueCommitment) -> Self::Output {
+        ValueCommitment((self.0 - rhs.0).into())
+    }
+}
+
+impl std::ops::SubAssign<ValueCommitment> for ValueCommitment {
+    fn sub_assign(&mut self, rhs: ValueCommitment) {
+        *self = *self - rhs;
+    }
+}
+
+impl std::iter::Sum for ValueCommitment {
+    fn sum<I>(iter: I) -> Self
+    where
+        I: Iterator<Item = Self>,
+    {
+        iter.fold(
+            ValueCommitment(pallas::Affine::identity()),
+            std::ops::Add::add,
+        )
+    }
+}
+
+/// LEBS2OSP256(repr_P(cv))
+///
+/// https://zips.z.cash/protocol/nu5.pdf#pallasandvesta
+impl TryFrom<[u8; 32]> for ValueCommitment {
+    type Error = &'static str;
+
+    fn try_from(bytes: [u8; 32]) -> Result<Self, Self::Error> {
+        let possible_point = pallas::Affine::from_bytes(&bytes);
+
+        if possible_point.is_some().into() {
+            Ok(Self(possible_point.unwrap()))
+        } else {
+            Err("Invalid pallas::Affine value")
+        }
+    }
+}
+
+impl ZcashSerialize for ValueCommitment {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer.write_all(&<[u8; 32]>::from(*self)[..])?;
+        Ok(())
+    }
+}
+
+impl ZcashDeserialize for ValueCommitment {
+    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        Self::try_from(reader.read_32_bytes()?).map_err(|e| SerializationError::Parse(e))
+    }
+}
+
+impl ValueCommitment {
+    /// Generate a new _ValueCommitment_.
+    ///
+    /// https://zips.z.cash/protocol/nu5.pdf#concretehomomorphiccommit
+    pub fn randomized<T>(csprng: &mut T, value: Amount) -> Self
+    where
+        T: RngCore + CryptoRng,
+    {
+        let rcv = generate_trapdoor(csprng);
+
+        Self::new(rcv, value)
+    }
+
+    /// Generate a new `ValueCommitment` from an existing `rcv on a `value`.
+    ///
+    /// ValueCommit^Orchard(v) :=
+    ///
+    /// https://zips.z.cash/protocol/nu5.pdf#concretehomomorphiccommit
+    #[allow(non_snake_case)]
+    pub fn new(rcv: pallas::Scalar, value: Amount) -> Self {
+        lazy_static! {
+            static ref V: pallas::Point = pallas_group_hash(b"z.cash:Orchard-cv", b"v");
+            static ref R: pallas::Point = pallas_group_hash(b"z.cash:Orchard-cv", b"r");
+        }
+
+        let v = pallas::Scalar::from(value);
+
+        Self::from(*V * v + *R * rcv)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Let's move the tests into an orchard::tests::{prop,vectors} module.

This can happen as a follow-up.